### PR TITLE
BUG: AttributeError: 'BooleanArray' object has no attribute 'sum' while infer types #44079

### LINKED ISSUE #44079: BUG: AttributeError: 'BooleanArray' object has no attribute 'sum' while infer types
Line 694 in the pandas/io/parsers/base_parser.py file, 
```
        if issubclass(values.dtype.type, (np.number, np.bool_)):
            # error: Argument 2 to "isin" has incompatible type "List[Any]"; expected
            # "Union[Union[ExtensionArray, ndarray], Index, Series]"
            mask = algorithms.isin(values, list(na_values))  # type: ignore[arg-type]
            na_count =mask.sum()
            ^^^^^^^^^^^^^THIS
            if na_count > 0:
                if is_integer_dtype(values):
                    values = values.astype(np.float64)
                np.putmask(values, mask, np.nan)
            return values, na_count
```

Caused an "AttributeError: 'BooleanArray' object has no attribute 'sum'" following some global ananconda update. Changed the line to 
`            na_count =mask.astype('uint8').sum()`

fixed the error in my case.

Edit by rhshadrach:

Minimal example

```
from io import StringIO

import pandas as pd

df = pd.read_csv(
     StringIO("0,1"),
     names=['a', 'b'],
     index_col=['a'],
     dtype={'a': 'UInt8'},
)
```

### PR DESCRIPTION
**Pull request by zhangxiaoxing**
- [x] closes #44079
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by zhangxiaoxing** [comment]
Well the failing check is a different problem. And its a future warning treated as error. I should be able to fix that but probably better on a different thread?

File: doc/source/user_guide/timeseries.rst [hunk 1]
```diff
@@ -2074,7 +2074,7 @@ The ``period`` dtype can be used in ``.astype(...)``. It allows one to change th
    pi.astype("period[D]")
 
    # convert to DatetimeIndex
-   pi.astype("datetime64[ns]")
```
Comment (jreback) [inline doc/source/user_guide/timeseries.rst:2077]: remove extraneous changes

File: pandas/io/parsers/base_parser.py [hunk 2]
```diff
@@ -700,7 +700,7 @@ def _infer_types(self, values, na_values, try_num_bool=True):
             # error: Argument 2 to "isin" has incompatible type "List[Any]"; expected
             # "Union[Union[ExtensionArray, ndarray], Index, Series]"
             mask = algorithms.isin(values, list(na_values))  # type: ignore[arg-type]
-            na_count = mask.sum()
+            na_count = mask.astype("uint8").sum()
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:703]: can add `copy=False`

File: pandas/tests/io/parser/test_index_col.py [hunk 3]
```diff
@@ -297,3 +297,24 @@ def test_multiindex_columns_index_col_with_data(all_parsers):
...
+    )
+    expected = DataFrame(
+        data={
+            "a": [
+                0,
+            ],
+            "b": [1],
+        }
+    ).set_index("a")
+    tm.assert_frame_equal(result, expected, check_index_type=False)
```
Comment (jreback) [inline pandas/tests/io/parser/test_index_col.py:320]: why is the check_index_type=False? can we not construct the appropriate expected?

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note in I/O section

File: pandas/tests/io/parser/test_index_col.py (hunk 3 above)
Comment (zhangxiaoxing) [reply pandas/tests/io/parser/test_index_col.py:320]: The C parser returns a index column of dtype 'object' and the Python parser returns a index column of dtype 'int64'. Both result are inferred, and both make sense in this specific example. Unifying the behavior of the parsers is desirable but beyond the scope of the current bug fix.

File: doc/source/user_guide/timeseries.rst (hunk 1 above)
Comment (zhangxiaoxing) [reply doc/source/user_guide/timeseries.rst]: done

File: pandas/io/parsers/base_parser.py (hunk 2 above)
Comment (zhangxiaoxing) [reply pandas/io/parsers/base_parser.py]: done

File: doc/source/whatsnew/v1.4.0.rst [hunk 4]
```diff
@@ -639,6 +639,7 @@ I/O
 - Bug in :func:`read_csv` with :code:`float_precision="round_trip"` which did not skip initial/trailing whitespace (:issue:`43713`)
 - Bug in dumping/loading a :class:`DataFrame` with ``yaml.dump(frame)`` (:issue:`42748`)
 - Bug in :func:`read_csv` raising ``ValueError`` when ``parse_dates`` was used with ``MultiIndex`` columns (:issue:`8991`)
+- Bug in :func:`read_csv` raising ``AttributeError`` when attempting to read .csv files and infer index column dtype (:issue:`44079`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:642]: this needs to specifically mention using an nullable integer type (as this is not generic rather a specific thing)

File: pandas/tests/io/parser/test_index_col.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/io/parser/test_index_col.py:320]: can you add a comment about why this is not checking index type (basically what you said here)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.4.0.rst (hunk 4 above)
Comment (zhangxiaoxing) [reply doc/source/whatsnew/v1.4.0.rst]: done

File: pandas/tests/io/parser/test_index_col.py (hunk 3 above)
Comment (zhangxiaoxing) [reply pandas/tests/io/parser/test_index_col.py:323]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @zhangxiaoxing

**Pull request merged**